target/linux/ipq806x/base-files/etc/init.d/hwmon_fancontrol

```diff
@@ -0,0 +1,20 @@
+#!/bin/sh /etc/rc.common
+
+START=98
+
+boot() {
+	local path_to_hwmon
+	# configuring onboard temp/fan controller to run the fan on its own
+	# for more information, please read https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface
+
+	case $(board_name) in
+	xiaomi,mi-router-hd)
+		path_to_hwmon="$( grep -l emc230 /sys/class/hwmon/hwmon*/name )"
+		if [ -n "$path_to_hwmon" ]; then
+			path_to_hwmon=$( dirname "$path_to_hwmon" 2>/dev/null )
+			# Set FAN speed to 80%
+			echo "204" > "$path_to_hwmon/pwm1"
```

Comment:
This is what is done on stock?

Reply:
On the stock firmware, the bash script is immediately launched, which controls the fan speed.
On OpenWRT firmware, the user must install additional software to control the fan. For this router I recommend this: https://github.com/openwrt-xiaomi/openwrt-alpine-fan-control/releases/latest

Reply:
Yes I mean what is the logic for making the fan turn on or off? Maybe we can describe that in DT with cooling-cell and trip points.

Reply:
I notice the thing have a temp sensor and the fan... So I guess everything is driven with the temp sensor... I think the 2 can be coupled and by defining some trip point you can describe the FAN function entirely in DT without using userspace scripts.

Reply:
Yes, this should be moved to use the in-kernel thermal trips and fan as the cooling device otherwise we will have a mess with various scripts as each device attempts to use different logic.

Reply:
> I think the 2 can be coupled and by defining some trip point you can describe the FAN function entirely in DT without using userspace scripts

It is better to do this as part of a separate pull request, since kernel patches are required.

Reply:
I will ask again:
> Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?

Reply:
> Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?

I didn't understand the question

Reply:
> > Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?
> 
> I didn't understand the question

What is registering the EMC230x device as the kernel driver for it does not support device tree at all.

Reply:
> Also, how is the EMC230x device even being registered as the current driver is not DT aware at all?

```
root@OpenWrt:~# grep -l -F emc230 /sys/class/hwmon/hwmon*/name
/sys/class/hwmon/hwmon11/name

root@OpenWrt:~# ls -la /sys/class/hwmon/hwmon11
lrwxrwxrwx    1 root     root  0 Sep 10 07:15 /sys/class/hwmon/hwmon11 -> ../../devices/platform/i2c_gpio_0/i2c-0/0-002f/hwmon/hwmon11

root@OpenWrt:~# ls -la /sys/devices/platform/i2c_gpio_0/i2c-0/0-002f/hwmon/hwmon11
drwxr-xr-x    2 root     root             0 Sep 10 07:15 .
drwxr-xr-x    3 root     root             0 Sep 10 07:15 ..
lrwxrwxrwx    1 root     root             0 Sep 10 07:15 device -> ../../../0-002f
-r--r--r--    1 root     root          4096 Sep 12 09:47 fan1_fault
-r--r--r--    1 root     root          4096 Sep 12 09:47 fan1_input
-r--r--r--    1 root     root          4096 Sep 10 07:15 name
lrwxrwxrwx    1 root     root             0 Sep 12 09:47 of_node -> ../../../../../../../firmware/devicetree/base/i2c_gpio_0/fan@2f
-rw-r--r--    1 root     root          4096 Sep 12 09:46 pwm1
lrwxrwxrwx    1 root     root             0 Sep 12 09:47 subsystem -> ../../../../../../../class/hwmon
-rw-r--r--    1 root     root          4096 Sep 12 09:47 uevent

root@OpenWrt:~# cat /sys/devices/platform/i2c_gpio_0/i2c-0/0-002f/hwmon/hwmon11/uevent
OF_NAME=fan
OF_FULLNAME=/i2c_gpio_0/fan@2f
OF_COMPATIBLE_0=microchip,emc2305
OF_COMPATIBLE_N=1
```